volume/drivers/coprhd/coprhd.go

```diff
@@ -138,42 +132,134 @@ func (d *driver) Create(
 	source *api.Source,
 	spec *api.VolumeSpec) (string, error) {
 
-	s, err := d.getAuthSession()
+	var err error
+
+	project := d.project
+	varray := d.varray
+	vpool := d.vpool
+	itr := d.itr
+
+	if name, ok := locator.VolumeLabels["project"]; ok {
+		project, err = d.coprhd.Project().
+			Name(name).
+			Query()
+		if err != nil {
+			return "", err
+		}
+	}
+	if project == nil {
+		return "", ErrProjectRequired
+	}
+
+	if name, ok := locator.VolumeLabels["varray"]; ok {
+		varray, err = d.coprhd.VArray().
+			Name(name).
+			Query()
+		if err != nil {
+			return "", err
+		}
+	}
+	if varray == nil {
+		return "", ErrArrayRequired
+	}
+
+	if name, ok := locator.VolumeLabels["vpool"]; ok {
+		vpool, err = d.coprhd.VPool().
+			Name(name).
+			Query()
+		if err != nil {
+			return "", err
+		}
+	}
+	if vpool == nil {
+		return "", ErrPoolRequired
+	}
+
+	if !vpool.IsBlock() {
+		return "", ErrInvalidPool
+	}
+
+	// make sure this pool supports the initiator protocol
+	found := false
+	for _, t := range vpool.Protocols {
+		if t == itr.Protocol {
+			found = true
+		}
+	}
+	if !found {
+		return "", ErrInvalidPort
+	}
+
+	sz := spec.Size
+	if sz < minVolumeSize {
+		sz = minVolumeSize
+	}
 
+	vol, err := d.coprhd.Volume().
+		Name(locator.Name).
+		Project(project.Id).
+		Array(varray.Id).
+		Pool(vpool.Id).
+		Create(sz)
 	if err != nil {
-		dlog.Errorf("Failed to create session: %s", err.Error())
 		return "", err
 	}
 
-	e := ApiError{}
+	dlog.Infof("coprhd volume %s created", vol.WWN)
 
-	res := &CreateVolumeReply{}
+	volumeID := strings.ToLower(vol.WWN)
 
-	sz := int64(spec.Size / (1024 * 1024 * 1000))
+	volume := common.NewVolume(
+		volumeID,
```

Comment:
i need to look through the other drivers, but generally it seems that volumeID is an osd-managed id, not an external id, the volume should probably account for this and make an id and then have the coprhd id be external @jvinod but this is debatable